### PR TITLE
Cleanup, improve directory structure

### PR DESCRIPTION
**Pull request by michaljurecko**
Jira: https://keboola.atlassian.net/browse/PSGO-286

Changes:
- Moved `buffer/storage/{local,staging,target}` -> `buffer/storage/`**`level`**`/{local,staging,target}`.
  - Because, I will add more sub-directories to the `storage`, and these levels are together a separated part of the implementation.
- Moved part of the `buffer/storage/local/writer` pkg to new `buffer/storage/local/writer/volume` pkg.
- Moved part of the `buffer/storage/local/reader` pkg to new `buffer/storage/local/reader/volume` pkg.

-----------------

image

---------------

Dal som do tohto PR radsej samostatne zmeny v adresarovej strukture, pred dalsimi PRs, aby toho nebolo vela.

### CODE REVIEW

File: internal/pkg/service/buffer/dependencies/dependencies.go [hunk 1]
```diff
@@ -37,8 +37,8 @@ import (
 	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/event"
 	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/file"
 	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/statistics"
-	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/storage/target/table"
-	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/storage/target/token"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/storage/level/target/table"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/storage/level/target/token"
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/dependencies/dependencies.go:41]: Do adresara `storage` pribudnu dalsie pod-adresare, preto som tam pridal novy adresar `level`, pre `{local,staging,target}` - co su rozne urovne storage, ktorymi data prejdu.

image

File: internal/pkg/service/buffer/storage/level/local/writer/csv/csv_test.go [hunk 2]
```diff
@@ -21,11 +21,10 @@ import (
 	"github.com/keboola/keboola-as-code/internal/pkg/log"
 	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/storage"
 	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/storage/compression"
-	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/storage/level/local/volume"
-	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/storage/level/local/writer"
 	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/storage/level/local/writer/csv"
 	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/storage/level/local/writer/disksync"
 	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/storage/level/local/writer/test/testcase"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/storage/level/local/writer/volume"
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/storage/level/local/writer/csv/csv_test.go]: `writer` adresar obsahoval uz toho privela + potreboval som to upravit aj aby boli prehladnejsie zavislosti medzi pkgs, 
(vacsiu) cast `writer` pkg som presunul do `writer/volume` pkg.

image

File: internal/pkg/service/buffer/storage/level/local/writer/factory.go [hunk 3]
```diff
@@ -7,9 +7,9 @@ import (
 	"github.com/keboola/keboola-as-code/internal/pkg/utils/errors"
 )
 
-type Factory func(w *base.Writer) (SliceWriter, error)
+type Factory func(w *base.Writer) (Writer, error)
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/storage/level/local/writer/factory.go:10]: `writer.SliceWriter` -> `writer.Writer` (iny writer tam nie je).

File: internal/pkg/service/buffer/storage/level/local/writer/csv/csv_test.go [hunk 4]
```diff
@@ -43,7 +43,7 @@ func TestCSVWriter_InvalidNumberOfValues(t *testing.T) {
 	defer cancel()
 
 	// Open volume
-	vol, err := volume.OpenVolume(ctx, log.NewNopLogger(), clock.New(), volume.NewInfo(t.TempDir(), "hdd", "1"))
+	vol, err := volume.Open(ctx, log.NewNopLogger(), clock.New(), volume.NewInfo(t.TempDir(), "hdd", "1"))
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/storage/level/local/writer/csv/csv_test.go]: `volume.OpenVolume` (predtym `writer.OpenVolume`) -> `volume.Open`.

File: internal/pkg/service/buffer/storage/level/local/reader/volumes.go [hunk 5]
```diff
@@ -16,7 +16,7 @@ type Volumes struct {
 }
 
 func OpenVolumes(ctx context.Context, logger log.Logger, clock clock.Clock, path string, opts ...Option) (*Volumes, error) {
-	v, err := volume.OpenVolumes(logger, path, func(info volumeInfo) (*Volume, error) {
+	v, err := volume.DetectVolumes(logger, path, func(info volumeInfo) (*Volume, error) {
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/storage/level/local/reader/volumes.go]: `volume.OpenVolumes` (predtym `writer.OpenVolumes`) -> `volume.DetectVolumes`:
 - Mame teda `volume.Open` a `volume.DetectVolumes`, nemalo by sa to vzajomne mylit.

File: internal/pkg/service/buffer/storage/level/local/writer/volume/volume.go [hunk 6]
```diff
@@ -243,18 +240,6 @@ func (v *Volume) checkDrainFile() error {
 	return nil
 }
 
-func (v *Volume) openedWriters() (out []writer.Writer) {
-	v.writersLock.Lock()
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/storage/level/local/writer/volume/volume.go]: Presunul som tuto metodu z `volume.go` do `writer.go`, kde je suvisiaci kod + je uz public ako `Writers`, budem to potrebovat v dalsom PR (pre ziskavanie statistik z jednotlivych writers).

File: internal/pkg/service/buffer/storage/level/local/writer/volume/drain.go [hunk 7]
```diff
@@ -0,0 +1,86 @@
+package volume
+
+import (
+	"github.com/fsnotify/fsnotify"
+	"github.com/keboola/keboola-as-code/internal/pkg/utils/errors"
+	"os"
+)
+
+func (v *Volume) Drained() bool {
+	return v.drained.Load()
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/storage/level/local/writer/volume/drain.go:12]: Vytvoreny subor `drain.go` a `drain_test.go`.

File: internal/pkg/service/buffer/storage/level/local/reader/volume/config.go [hunk 8]
```diff
@@ -1,4 +1,4 @@
-package reader
+package volume
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/storage/level/local/reader/volume/config.go:1]: V nasledujucich commits je to iste co bolo riesene doteraz pre `writer`, aj pre `reader`.

V tomto commite je presunutie kodu do `volume` pkg.

image

**Review by zajca — APPROVED** [review]

**Pull request merged**